File: apps/hiveweave/lib/hiveweave/services/team_chat.ex
defmodule HiveWeave.Services.TeamChat do
  @moduledoc """
  TeamChatService — multi-agent group chat persistence and deduplication.

  Wraps ChatMessageService to record inter-agent team communications
  as 'team' role messages visible in the UI team comms panel.
  """

  alias HiveWeave.Repo.ProjectFactory

  require Logger

  @doc """
  Record a team chat message between agents.
  Deduplicates by (from, to, dedupe_key) within a time window.
  """
  def record_message(agent_id, from_agent_id, to_agent_id, content, _opts \\ %{}) do
    dedupe_key = build_dedupe_key(from_agent_id, to_agent_id, content)

    if is_duplicate?(agent_id, dedupe_key) do
      :duplicate
    else
      now_ms = System.system_time(:millisecond)
      id = Ecto.UUID.generate()

      attrs = %{
        id: id,
        agent_id: agent_id,
        role: "team",
        content: content,
        is_background: false,
        is_read: false,
        is_streaming: false,
        created_at: now_ms,
        team_from_agent_id: from_agent_id,
        team_to_agent_id: to_agent_id
      }

      case save_team_message(attrs) do
        {:ok, _} ->
          # Save dedupe key
          save_dedupe_key(agent_id, dedupe_key, now_ms)
          :ok

        {:error, reason} ->
          Logger.warning("[TeamChat] Failed to record message: #{inspect(reason)}")
          {:error, reason}
      end
    end
  end

  @doc """
  Get team chat history for a project/agent.
  """
  def get_history(agent_id, limit \\ 50) do
    sql = """
    SELECT id, agent_id, content, team_from_agent_id, team_to_agent_id, created_at
    FROM chat_messages
    WHERE role = 'team' AND agent_id = ?
    ORDER BY created_at DESC
    LIMIT ?
    """

    case ProjectFactory.query_for_agent(agent_id, sql, [agent_id, limit]) do
      {:ok, r} ->
        r.rows
        |> Enum.reverse()
        |> Enum.map(fn [id, aid, content, from_id, to_id, created] ->
          %{
            id: id, agent_id: aid, content: content,
            from_agent_id: from_id, to_agent_id: to_id,
            created_at: created
          }
        end)

      {:error, _} -> []
    end
  end

  # ── Save team message with team columns ────────────────────────

  defp save_team_message(attrs) do
    id = Map.get(attrs, :id)
    agent_id = Map.get(attrs, :agent_id)
    role = Map.get(attrs, :role, "team")
    content = Map.get(attrs, :content, "")
    is_background = if Map.get(attrs, :is_background, false), do: 1, else: 0
    is_read = if Map.get(attrs, :is_read, true), do: 1, else: 0
    is_streaming = if Map.get(attrs, :is_streaming, false), do: 1, else: 0
    created_at = Map.get(attrs, :created_at, System.system_time(:millisecond))
    team_from = Map.get(attrs, :team_from_agent_id, "")
    team_to = Map.get(attrs, :team_to_agent_id, "")

    sql = """
    INSERT INTO chat_messages (id, agent_id, role, content, tool_calls, is_background, is_read, is_streaming, team_from_agent_id, team_to_agent_id, created_at)
    VALUES (?, ?, ?, ?, '[]', ?, ?, ?, ?, ?, ?)
    """

    case ProjectFactory.query_for_agent(agent_id, sql, [
           id, agent_id, role, content, is_background, is_read, is_streaming, team_from, team_to, created_at
         ]) do
      {:ok, _} ->
        {:ok, %{id: id, role: role, content: content, created_at: created_at}}

      {:error, reason} ->
        {:error, reason}
    end
  rescue
    e -> {:error, e}
  end

  # ── Deduplication ──────────────────────────────────────

  @dedupe_window_ms 60_000  # 1 minute dedup window

  defp build_dedupe_key(from_id, to_id, content) do
    hash = :crypto.hash(:md5, "#{from_id}:#{to_id}:#{content}")
    Base.encode16(hash, case: :lower)
  end

  defp is_duplicate?(agent_id, key) do
    cutoff = System.system_time(:millisecond) - @dedupe_window_ms

    case ProjectFactory.query_for_agent(agent_id,
           "SELECT id FROM team_chat_dedupe WHERE agent_id = ? AND dedupe_key = ? AND created_at > ?",
           [agent_id, key, cutoff]) do
      {:ok, r} -> r.rows != []
      _ -> false
    end
  rescue
    _ -> false
  end

  defp save_dedupe_key(agent_id, key, now_ms) do
    id = Ecto.UUID.generate()
    ProjectFactory.query_for_agent(agent_id,
      "INSERT INTO team_chat_dedupe (id, agent_id, dedupe_key, created_at) VALUES (?, ?, ?, ?)",
      [id, agent_id, key, now_ms])
  rescue
    _ -> :ok
  end
end

File: apps/hiveweave-py/src/hiveweave/services/team_chat.py
"""Team chat service — multi-agent group chat.

契约 18: TeamChatService
- 复用 chat_messages 表（role='team'），独立 team_chat_dedupe 去重表
- 1 分钟窗口内 (from, to, content) 三元组重复则丢弃（返回 'duplicate'）
- record_message: 去重后写 chat_messages + dedupe_key
- get_history: DESC + reverse → 正序（oldest first）

team_chat_dedupe 表 schema 已存在于 per-project DB，无需迁移。
"""

import hashlib
import time
import uuid

import structlog

from hiveweave.db import project as project_db

log = structlog.get_logger(__name__)

# 契约 18: 去重窗口 60_000 ms（1 分钟）
_DEDUPE_WINDOW_MS = 60_000


class TeamChatService:
    """Multi-agent team chat with dedup.

    所有操作路由到 per-project DB（通过 agent_id）。
    消息存储在 chat_messages 表（role='team'），用 team_from/to_agent_id 区分。
    """

    async def record_message(self, agent_id: str, from_agent_id: str,
                             to_agent_id: str, content: str,
                             opts: dict | None = None) -> str:
        """Record a team chat message with dedup.

        契约 18: record_message
        - dedupe_key = MD5("{from}:{to}:{content}") hex lowercase
        - 1 分钟窗口内重复 → 返回 'duplicate'，不写 chat_messages
        - 非重复 → 写 chat_messages + dedupe_key，返回 'ok'
        - save 失败 → 返回 'error'

        TEST16 P1-2: dedupe 改用 INSERT OR IGNORE + UNIQUE 约束原子化，
        消除 check-then-act TOCTOU 竞态。

        Returns: 'ok' | 'duplicate' | 'error'
        """
        dedupe_key = hashlib.md5(
            f"{from_agent_id}:{to_agent_id}:{content}".encode()
        ).hexdigest()
        now_ms = int(time.time() * 1000)
        window_bucket = now_ms // _DEDUPE_WINDOW_MS

        # Atomic dedupe claim (fail-open: exception → not duplicate)
        dedupe_id = str(uuid.uuid4())
        try:
            await project_db.execute(
                agent_id,
                "INSERT OR IGNORE INTO team_chat_dedupe "
                "(id, agent_id, dedupe_key, created_at, window_bucket) "
                "VALUES (?, ?, ?, ?, ?)",
                [dedupe_id, agent_id, dedupe_key, now_ms, window_bucket])
            row = await project_db.query_one(
                agent_id,
                "SELECT id FROM team_chat_dedupe WHERE id = ?",
                [dedupe_id])
            if row is None:
                log.info("team_chat_dedup", agent_id=agent_id,
                         from_agent_id=from_agent_id, to_agent_id=to_agent_id)
                return "duplicate"
        except Exception as e:
            log.warning("team_chat_dedupe_claim_failed", agent_id=agent_id,
                        error=str(e))
            # fail-open: proceed with message save

        # Save message to chat_messages
        msg_id = str(uuid.uuid4())
        try:
            await project_db.execute(
                agent_id,
                "INSERT INTO chat_messages (id, agent_id, role, content, "
                "is_background, is_read, is_streaming, team_from_agent_id, "
                "team_to_agent_id, created_at) "
                "VALUES (?, ?, 'team', ?, 0, 0, 0, ?, ?, ?)",
                [msg_id, agent_id, content, from_agent_id, to_agent_id, now_ms])
        except Exception as e:
            log.error("team_chat_save_failed", agent_id=agent_id, error=str(e))
            return "error"

        return "ok"

    async def check_and_mark(self, agent_id: str, from_agent_id: str,
                             to_agent_id: str, content: str) -> bool:
        """原子化"检查+登记"去重（供不写 chat_messages 的调用方使用）。

        TEST16 P1-2: 改用 INSERT OR IGNORE + UNIQUE(agent_id, dedupe_key,
        window_bucket) 约束兜底并发。不再依赖 check-then-act 的 SELECT+INSERT
        两步操作（存在 TOCTOU 竞态）。

        Returns: True = duplicate（跳过写库）; False = newly marked（可写库）
        """
        dedupe_key = hashlib.md5(
            f"{from_agent_id}:{to_agent_id}:{content}".encode()
        ).hexdigest()
        now_ms = int(time.time() * 1000)
        window_bucket = now_ms // _DEDUPE_WINDOW_MS
        dedupe_id = str(uuid.uuid4())
        try:
            await project_db.execute(
                agent_id,
                "INSERT OR IGNORE INTO team_chat_dedupe "
                "(id, agent_id, dedupe_key, created_at, window_bucket) "
                "VALUES (?, ?, ?, ?, ?)",
                [dedupe_id, agent_id, dedupe_key, now_ms, window_bucket])
            # Check if OUR row was inserted (we won the race)
            row = await project_db.query_one(
                agent_id,
                "SELECT id FROM team_chat_dedupe WHERE id = ?",
                [dedupe_id])
            if row is not None:
                return False  # newly marked — caller may proceed
            # Our INSERT was ignored → duplicate within this window bucket
            log.info("team_chat_dedup_mark", agent_id=agent_id,
                     from_agent_id=from_agent_id, to_agent_id=to_agent_id)
            return True
        except Exception as e:
            log.warning("team_chat_check_mark_failed", agent_id=agent_id,
                        error=str(e))
            return False

    async def get_history(self, agent_id: str, limit: int = 50) -> list[dict]:
        """Get team chat history (chronological order). Default limit 50.

        契约 18: get_history
        - WHERE role='team' AND agent_id=?
        - ORDER BY created_at DESC LIMIT ? → reverse → 正序（oldest first）
        - 返回 [{id, agent_id, content, from_agent_id, to_agent_id, created_at}]
        - 异常返回 []
        """
        try:
            rows = await project_db.query(
                agent_id,
                "SELECT id, agent_id, content, team_from_agent_id, "
                "team_to_agent_id, created_at "
                "FROM chat_messages WHERE role = 'team' AND agent_id = ? "
                "ORDER BY created_at DESC LIMIT ?",
                [agent_id, limit])
            return [self._row_to_msg(r) for r in reversed(rows)]
        except Exception as e:
            log.warning("team_chat_history_failed", agent_id=agent_id,
                        error=str(e))
            return []

    async def _is_duplicate(self, agent_id: str, dedupe_key: str,
                            cutoff: int) -> bool:
        """Check if a dedupe_key exists within the window.

        契约 18: is_duplicate? — 异常返回 False（fail-open，宁可重复不丢消息）。
        """
        try:
            row = await project_db.query_one(
                agent_id,
                "SELECT id FROM team_chat_dedupe "
                "WHERE agent_id = ? AND dedupe_key = ? AND created_at > ? LIMIT 1",
                [agent_id, dedupe_key, cutoff])
            return row is not None
        except Exception as e:
            log.warning("team_chat_dedup_check_failed", agent_id=agent_id,
                        error=str(e))
            return False

    @staticmethod
    def _row_to_msg(row) -> dict:
        """Convert DB row to message dict with friendly key names."""
        d = dict(row)
        d["from_agent_id"] = d.pop("team_from_agent_id", None)
        d["to_agent_id"] = d.pop("team_to_agent_id", None)
        return d
